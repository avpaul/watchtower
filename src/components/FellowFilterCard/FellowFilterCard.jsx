import React from 'react';
import PropTypes from 'prop-types';
import check from '../../static/check-mark.svg';
import defaultPicture from '../../static/profile.svg';

/**
 * @method FellowFilterCard
 * @summary - Method to render fellow's cards that can be filtered
 * @param { cardName {string},
 * numberOfFellows { number }, isTicked {object - Project|status},
 * handleClick { function }, filterKey {string} } param0 {Object}
 * @returns A rendered version of a clickable fellows card with the name, number of fellows,
 * whether the card should be ticked and a key to be filtered by.
 */
const FellowFilterCard = ({
  cardName,
  numberOfFellows,
  isTicked,
  handleCardClick,
  filterKey,
  isManager,
  picture = '',
  displayPicture
}) => {
  const shouldDisplayTick =
    isTicked.project === cardName ||
    isTicked.status === cardName ||
    isTicked.type === cardName ||
    isTicked.criteria === cardName ||
    isTicked.level === cardName ||
    isTicked.manager_email === cardName;

  const shouldRenderImage =
    isManager && displayPicture ? (
      <img className="manager-picture" src={picture || defaultPicture} alt="" />
    ) : (
      ''
    );

  const adjustCardNameStyles = isManager && displayPicture ? `pl-3` : '';
  const cardNameStyle = isManager
    ? 'manager-filter-card'
    : 'fellow-filter-card';

  const styleCardNumber = `float-right right ${
    isManager ? 'manager-card__top mb-2' : 'mt-3'
  }  card-number-display`;

  return (
    <div
      className={`fellow-summary-card ${cardNameStyle} mt-4
    ${shouldDisplayTick ? 'fellow-card-highlight' : ''}`}
      id={cardName}
      filterKey={filterKey}
      onClick={handleCardClick}
      onKeyPress={handleCardClick}
      role="button"
      tabIndex="0"
    >
      <div className="">
        <img
          className="img-responsive rounded-circle checkmark float-right"
          style={{
            display: shouldDisplayTick ? 'block' : 'none'
          }}
          src={check}
          alt=""
        />
      </div>
      <div className="float-left left-card mt-3 card-name">
        <div className="row">
          <div className="">{shouldRenderImage}</div>
          <div className="">
            <span className={adjustCardNameStyles}>{cardName}</span>
          </div>
        </div>
      </div>
      <div className={styleCardNumber}>{numberOfFellows}</div>
    </div>
  );
};

/**
 * @name FellowFilterCard Proptypes
 * @type (Proptypes)
 * @property cardName - String
 * @property numberOfFellow - number
 * @property handleCardClick - function
 * @property filterKey - string
 */
FellowFilterCard.propTypes = {
  cardName: PropTypes.string.isRequired,
  numberOfFellows: PropTypes.number.isRequired,
  isTicked: PropTypes.shape({}).isRequired,
  handleCardClick: PropTypes.func.isRequired,
  filterKey: PropTypes.string.isRequired,
  isManager: PropTypes.bool,
  picture: PropTypes.string,
  displayPicture: PropTypes.bool
};

FellowFilterCard.defaultProps = {
  picture: '',
  isManager: false,
  displayPicture: false
};

export default FellowFilterCard;
