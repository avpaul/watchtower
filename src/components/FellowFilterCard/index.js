import FellowFilterCard from './FellowFilterCard';

export default FellowFilterCard;
