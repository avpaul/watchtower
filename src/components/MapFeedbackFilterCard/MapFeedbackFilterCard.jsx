import React from 'react';
import PropTypes from 'prop-types';
import FellowFilterCard from '../FellowFilterCard';
import TranslatorTable from '../../utils/TranslatorTable';
import { processArray, processCountInformation } from '../../services/helper';

const MapFeedbackFilterCard = ({
  feedbackArray,
  title,
  filterKey,
  isTicked = {},
  handleCardClick
}) => {
  let refinedArray = [{ [title]: feedbackArray.length || 0 }];
  const feedbackFilters = {};

  const processedResult = processCountInformation(
    feedbackArray,
    feedbackFilters,
    filterKey
  );

  refinedArray = processArray(refinedArray, processedResult);

  const getCardName = fellows =>
    TranslatorTable[Object.keys(fellows)[0]]
      ? TranslatorTable[Object.keys(fellows)[0]]
      : Object.keys(fellows)[0];

  return (
    <div className="ops-dashboard__fellows-summary">
      <div className="row">
        {refinedArray.map(fellows => (
          <FellowFilterCard
            cardName={getCardName(fellows)}
            numberOfFellows={fellows[Object.keys(fellows)[0]]}
            isTicked={isTicked}
            handleCardClick={handleCardClick}
            filterKey={filterKey}
          />
        ))}
      </div>
    </div>
  );
};

MapFeedbackFilterCard.propTypes = {
  feedbackArray: PropTypes.shape([]).isRequired,
  title: PropTypes.string.isRequired,
  filterKey: PropTypes.string.isRequired,
  isTicked: PropTypes.shape({}).isRequired,
  handleCardClick: PropTypes.func.isRequired
};

export default MapFeedbackFilterCard;
