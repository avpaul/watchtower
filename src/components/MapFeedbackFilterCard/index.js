import MapFeedbackFilterCard from './MapFeedbackFilterCard';

export default MapFeedbackFilterCard;
