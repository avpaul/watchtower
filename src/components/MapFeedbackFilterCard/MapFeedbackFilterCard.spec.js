import React from 'react';
import { shallow } from 'enzyme';
import feedbackData from '../../__mocks__/feedbackSummary.json';
import MapFeedbackFilterCard from './MapFeedbackFilterCard';
import FellowFilterCard from '../FellowFilterCard';

describe('MapFeedbackFilter card tests', () => {
  const setup = (
    feedbackArray,
    cardTitle,
    useFilterData,
    filterKey,
    isTicked
  ) => {
    const props = {
      feedbackArray,
      filteredFeedbackArray: [],
      useFilterData,
      title: cardTitle,
      filterKey,
      isTicked,
      handleCardClick: jest.fn()
    };
    const MapFeedbackFilterCardWrapper = shallow(
      <MapFeedbackFilterCard {...props} />
    );
    return { MapFeedbackFilterCardWrapper, props };
  };

  it('should render MapFeedbackFilter card without crashing', () => {
    const { MapFeedbackFilterCardWrapper } = setup(
      feedbackData,
      'All Products',
      false,
      'project',
      { project: 'All Projects' }
    );
    expect(MapFeedbackFilterCardWrapper).toMatchSnapshot();
  });

  it('should render 3 filter cards for all products', () => {
    const { MapFeedbackFilterCardWrapper } = setup(
      feedbackData,
      'All Products',
      false,
      'project',
      { project: 'All Projects' }
    );
    expect(MapFeedbackFilterCardWrapper.find(FellowFilterCard).length).toBe(3);
  });
  it('should render the filter cards below the slider apporpriately', () => {
    const { MapFeedbackFilterCardWrapper } = setup(
      feedbackData,
      'All Criteria',
      false,
      'pre-pip',
      { type: 'Pre-PIP only' }
    );
    expect(
      MapFeedbackFilterCardWrapper.find(FellowFilterCard)
        .at(1)
        .prop('numberOfFellows')
    ).toEqual(4);
  });
});
