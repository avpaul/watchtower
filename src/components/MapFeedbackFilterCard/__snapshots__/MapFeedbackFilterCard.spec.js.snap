// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapFeedbackFilter card tests should render MapFeedbackFilter card without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MapFeedbackFilterCard
    feedbackArray={
        Array [
            Object {
              "context": "Your quality is below expectation, you are off-track",
              "criteria": "lms",
              "end_date": "null",
              "first_name": "Oluseyi",
              "last_name": "Anani",
              "level": "D0B Apprenticeship",
              "manager": "Trust birungi",
              "manager_email": "trust.birungi@andela.com",
              "project": "Watch Tower",
              "start_date": "2019-03-05",
              "type": "pre-pip",
            },
            Object {
              "context": "Your quality is below expectation",
              "criteria": "pulse",
              "end_date": "2018-07-18",
              "first_name": "Brian",
              "last_name": "Mboya",
              "level": "D0A Simulations",
              "manager": "Trust birungi",
              "manager_email": "trust.birungi@andela.com",
              "project": "Noldor",
              "start_date": "2019-03-06",
              "type": "pip",
            },
            Object {
              "context": "Your quality is below expectation",
              "criteria": "pulse",
              "end_date": "2018-07-18",
              "first_name": "Brian",
              "last_name": "Mboya",
              "level": "D0A Simulations",
              "manager": "Trust birungi",
              "manager_email": "trust.birungi@andela.com",
              "project": "Noldor",
              "start_date": "2019-03-08",
              "type": "pip",
            },
            Object {
              "context": "Your quality is below expectation",
              "criteria": "lms",
              "end_date": "2018-07-18",
              "first_name": "Brian",
              "last_name": "Mboya",
              "level": "D0A Simulations",
              "manager": "Trust birungi",
              "manager_email": "trust.birungi@andela.com",
              "project": "Noldor",
              "start_date": "2019-03-10",
              "type": "pip",
            },
          ]
    }
    filterKey="project"
    handleCardClick={[Function]}
    isTicked={
        Object {
            "project": "",
          }
    }
    title="All Products"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <FellowFilterCard
          cardName="All Products"
          displayPicture={true}
          filterKey="project"
          handleCardClick={[Function]}
          isManager={false}
          isTicked={
                    Object {
                              "project": "",
                            }
          }
          numberOfFellows={4}
          picture=""
/>,
        <FellowFilterCard
          cardName="Watch Tower"
          displayPicture={true}
          filterKey="project"
          handleCardClick={[Function]}
          isManager={false}
          isTicked={
                    Object {
                              "project": "",
                            }
          }
          numberOfFellows={1}
          picture=""
/>,
        <FellowFilterCard
          cardName="Noldor"
          displayPicture={true}
          filterKey="project"
          handleCardClick={[Function]}
          isManager={false}
          isTicked={
                    Object {
                              "project": "",
                            }
          }
          numberOfFellows={3}
          picture=""
/>,
      ],
      "className": "manager-slick",
      "infinite": false,
      "nextArrow": <NavArrow
        buttonClass="slick-next"
        iconClass="fa-angle-right"
        onClick={[Function]}
/>,
      "prevArrow": <NavArrow
        buttonClass="slick-prev"
        iconClass="fa-angle-left"
        onClick={[Function]}
/>,
      "responsive": Array [
        Object {
          "breakpoint": 1000,
          "settings": Object {
            "slidesToScroll": 1,
            "slidesToShow": 2,
          },
        },
        Object {
          "breakpoint": 700,
          "settings": Object {
            "slidesToScroll": 1,
            "slidesToShow": 1,
          },
        },
      ],
      "rows": 1,
      "slidesToShow": 2.9999,
      "swipeToSlide": true,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "All Products",
        "nodeType": "function",
        "props": Object {
          "cardName": "All Products",
          "displayPicture": true,
          "filterKey": "project",
          "handleCardClick": [Function],
          "isManager": false,
          "isTicked": Object {
            "project": "",
          },
          "numberOfFellows": 4,
          "picture": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "Watch Tower",
        "nodeType": "function",
        "props": Object {
          "cardName": "Watch Tower",
          "displayPicture": true,
          "filterKey": "project",
          "handleCardClick": [Function],
          "isManager": false,
          "isTicked": Object {
            "project": "",
          },
          "numberOfFellows": 1,
          "picture": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "Noldor",
        "nodeType": "function",
        "props": Object {
          "cardName": "Noldor",
          "displayPicture": true,
          "filterKey": "project",
          "handleCardClick": [Function],
          "isManager": false,
          "isTicked": Object {
            "project": "",
          },
          "numberOfFellows": 3,
          "picture": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <FellowFilterCard
            cardName="All Products"
            displayPicture={true}
            filterKey="project"
            handleCardClick={[Function]}
            isManager={false}
            isTicked={
                        Object {
                                    "project": "",
                                  }
            }
            numberOfFellows={4}
            picture=""
/>,
          <FellowFilterCard
            cardName="Watch Tower"
            displayPicture={true}
            filterKey="project"
            handleCardClick={[Function]}
            isManager={false}
            isTicked={
                        Object {
                                    "project": "",
                                  }
            }
            numberOfFellows={1}
            picture=""
/>,
          <FellowFilterCard
            cardName="Noldor"
            displayPicture={true}
            filterKey="project"
            handleCardClick={[Function]}
            isManager={false}
            isTicked={
                        Object {
                                    "project": "",
                                  }
            }
            numberOfFellows={3}
            picture=""
/>,
        ],
        "className": "manager-slick",
        "infinite": false,
        "nextArrow": <NavArrow
          buttonClass="slick-next"
          iconClass="fa-angle-right"
          onClick={[Function]}
/>,
        "prevArrow": <NavArrow
          buttonClass="slick-prev"
          iconClass="fa-angle-left"
          onClick={[Function]}
/>,
        "responsive": Array [
          Object {
            "breakpoint": 1000,
            "settings": Object {
              "slidesToScroll": 1,
              "slidesToShow": 2,
            },
          },
          Object {
            "breakpoint": 700,
            "settings": Object {
              "slidesToScroll": 1,
              "slidesToShow": 1,
            },
          },
        ],
        "rows": 1,
        "slidesToShow": 2.9999,
        "swipeToSlide": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "All Products",
          "nodeType": "function",
          "props": Object {
            "cardName": "All Products",
            "displayPicture": true,
            "filterKey": "project",
            "handleCardClick": [Function],
            "isManager": false,
            "isTicked": Object {
              "project": "",
            },
            "numberOfFellows": 4,
            "picture": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "Watch Tower",
          "nodeType": "function",
          "props": Object {
            "cardName": "Watch Tower",
            "displayPicture": true,
            "filterKey": "project",
            "handleCardClick": [Function],
            "isManager": false,
            "isTicked": Object {
              "project": "",
            },
            "numberOfFellows": 1,
            "picture": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "Noldor",
          "nodeType": "function",
          "props": Object {
            "cardName": "Noldor",
            "displayPicture": true,
            "filterKey": "project",
            "handleCardClick": [Function],
            "isManager": false,
            "isTicked": Object {
              "project": "",
            },
            "numberOfFellows": 3,
            "picture": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
